include/circt/Dialect/FIRRTL/FIRRTLInterfaces.td

```diff
@@ -0,0 +1,70 @@
+//===- FIRRTLInterfaces.td - FIRRTL op interfaces ----------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This describes the MLIR interfaces for FIRRTL operations.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FIRRTL_INTERFACES_TD
+#define FIRRTL_INTERFACES_TD
+
+include "mlir/IR/OpBase.td"
+
+// Op result type is entirely defined by its arguments.
+def ComputedType : OpInterface<"ComputedType"> {
+  let description = [{
+    An interface to operations whose result type is entirely defined by their
+    arguments (operand types and attributes).
+  }];
+  let methods = [
+
+    InterfaceMethod<[{
+      This method computes the result type of the operation from its arguments.
+      Operations must implement a `getResultType` function which actually
+      determines the type:
+
+      ```c++
+      class ConcreteOp ... {
+      protected:
+        FIRRTLType getResultType() { ... }
+      };
+      ```
+    }], "FIRRTLType", "getResultType", (ins), [{
+      return $_op.getResultType();
+    }]>,
+
+    InterfaceMethod<[{
+      This method changes the operation's result type to match the type returned
+      by `getResultType`. Returns true if the type changed.
+    }], "bool", "updateResultType", (ins), [{
+      auto newType = $_op.getResultType();
+      if (!newType)
+        return false;
+      if ($_op.getType() == newType)
+        return false;
+      $_op.getResult().setType(newType);
+      return true;
+    }]>,
+
+    InterfaceMethod<[{
+      Verify that the operation's result type matches the type determined from
+      its arguments.
+    }], "::mlir::LogicalResult", "verifyResultType", (ins), [{}], [{
+      auto resType = $_op.getResultType();
+      if (!resType)
+        return failure(); // already diagnosed the error
+      if (resType != $_op.getType())
```

Comment:
This seems like it depends on the Op having a single result.  Is there something we can do to ensure that the op interface is only on a single result op, or make this more generic?

Reply:
Yeah that would be great. Currently adding the interface on an op with multiple results will produce a bunch of compile-time errors. Making this more generic would be nice, but I'm not sure if we should do this here (despite not needing it at the moment in FIRRTL), or rather try to contribute something more general back upstream later 🤔.

Reply:
All FIRRTL primitives are currently guaranteed to have a single result type, we assume that in a bunch of firrtl specific code.  Instances and mems have multiple results tho

Reply:
Yeah good point. These are likely to want to recompute their result types too later on, once we do inter-module width inference. Might be worthwhile to keep that in mind.